src/routes/write-exercise/components/WriteExercise.tsx

```diff
@@ -130,23 +127,49 @@ const WriteExercise = ({
   const giveUp = async (): Promise<void> => {
     const previousResult = result
     setResult('incorrect')
-    storeResult(previousResult ?? 'incorrect')
+    storeResult(previousResult ?? 'incorrect', true)
   }
 
   const continueExercise = (): void => {
     setResult(null)
     setSubmission(null)
     setInput('')
 
-    if (currentDocumentNumber === documents.length - 1) {
+    if (currentDocumentNumber === documents.length - 1 && !getRetryInfoOfCurrent().needsToBeRepeated) {
       finishExercise(results)
     } else {
-      setCurrentDocumentNumber(currentDocumentNumber + 1)
+      getRetryInfoOfCurrent().needsToBeRepeated ? tryLater() : setCurrentDocumentNumber(currentDocumentNumber + 1)
     }
   }
 
-  const storeResult = (score: SimpleResultType): void => {
-    setResults([...results, { ...document, result: score }])
+  const storeResult = (score: SimpleResultType, noFurtherRetries: boolean = false): void => {
+    const nthRetry = noFurtherRetries ? 3 : getRetryInfoOfCurrent().nthRetry
+
+    const indexOfCurrentResult = results.findIndex(result => result.id === document?.id)
```

Comment:
Just use `find` and `map` instead

Reply:
Same as above, using map for adding a new entry does not work. Also we need to copy the array so useState notices the update, just mutating the array is not recommended as changes can be overwritten.

Reply:
Hmmm yeah. I think the code is really complicated and hard to understand though. Some ways I could think that perhaps could make it easier:
* Use a map (`{ [id]: result }`) instead of an array to store the results
* Initialize the results array for all documents

Reply:
I though about this a little and i think it makes sense to split `WriteExercise` up into two components.
One holds knows all the stuff relevant for all words of the exercise, e.g. which is the currentWord, what are all the results and documents, how may words are in this exercise
One knows only stuff about the current word: Which is the current word, what was submitted and so on.

Are you fine with this? If so, I will do it in a separate PR.

Reply:
Sounds really reasonable, thanks!